Sources/StreamChat/Workers/Background/MessageSender.swift

```diff
@@ -184,38 +191,66 @@ private class MessageSendingQueue {
-            // switch to using a custom `OrderedSet`
-            guard let request = self?.requests.sorted(by: { $0.createdLocallyAt < $1.createdLocallyAt }).first else { return }
-
-            self?.messageRepository.sendMessage(with: request.messageId) { [weak self] result in
-                guard let self else { return }
-                self.removeRequestAndContinue(request)
-                if let error = result.error {
-                    switch error {
-                    case .messageDoesNotExist,
-                         .messageNotPendingSend,
-                         .messageDoesNotHaveValidChannel:
-                        let event = NewMessageErrorEvent(messageId: request.messageId, error: error)
-                        self.eventsNotificationCenter.process(event)
-                    case let .failedToSendMessage(error):
-                        let event = NewMessageErrorEvent(messageId: request.messageId, error: error)
-                        self.eventsNotificationCenter.process(event)
-                    }
+            guard let self else { return }
+            guard let request = self.sortedQueuedRequests.first else { return }
+
+            if self.isWaitingForConnection {
+                self.messageRepository.scheduleOfflineRetry(for: request.messageId) { [weak self] _ in
+                    self?._requests.mutate { $0.remove(request) }
+                    self?.sendNextMessage()
+                }
+            } else {
+                self.messageRepository.sendMessage(with: request.messageId) { [weak self] result in
+                    self?.handleSendMessageResult(request, result: result)
                 }
-                self.delegate?.messageSendingQueue(self, didProcess: request.messageId, result: result)
             }
         }
     }
-
-    private func removeRequestAndContinue(_ request: SendRequest) {
+    
+    private func handleSendMessageResult(_ request: SendRequest, result: Result<ChatMessage, MessageRepositoryError>) {
         _requests.mutate { $0.remove(request) }
-        sendNextMessage()
+        
+        if let repositoryError = result.error {
+            switch repositoryError {
+            case .messageDoesNotExist, .messageNotPendingSend, .messageDoesNotHaveValidChannel:
+                let event = NewMessageErrorEvent(messageId: request.messageId, error: repositoryError)
+                eventsNotificationCenter.process(event)
+            case .failedToSendMessage(let clientError):
+                let event = NewMessageErrorEvent(messageId: request.messageId, error: clientError)
+                eventsNotificationCenter.process(event)
+                
+                if ClientError.isEphemeral(error: clientError) {
+                    // We hit a connection error, therefore all the remaining and upcoming requests should be scheduled for keeping the order
+                    isWaitingForConnection = true
```

Comment:
I see we only set this value to true here, shouldn't we always set it to true the `didUpdateConnectionStatus` is called? 🤔

Reply:
Because we have automatic retries in APIClient then it is better to give APIClient a chance to try to send it. Message sender sends messages serially so it is not a problem to do it later here when the request actually failed because of connection.